---
title: "Bimm 143 Class 12"
author: "Anika Bhattacharjya (PID: A15459876)"
output: html_document
---

# Comparative Analysis of Protein Structures
```{r}
library(bio3d)

pdb <- read.pdb("1hel")
pdb
```

Let's use bioinformatics methods called NMA (Normal Mode Analysis) to predict the dynamics (flexibility) of the enzyme

```{r}
modes <- nma(pdb)
plot(modes)
```

Make a "move" of its predicted motion. We often call this a "trajectory"

```{r}
mktrj(modes, file="nma.pdb")
```
Upload image of nma.pdb
![](nmapdb.png)
# Analysis of ADK
```{r}
aa <- get.seq("1ake_A")

```
```{r}
# List out some 'top hits'
hits <- NULL
hits$pdb.id <- c('1AKE_A','4X8M_A','6S36_A','6RZE_A','4X8H_A','3HPR_A','1E4V_A','5EJE_A','1E4Y_A','3X2S_A','6HAP_A','6HAM_A','4K46_A','4NP6_A','3GMT_A','4PZL_A')
```

```{r}
# Download releated PDB files
files <- get.pdb(hits$pdb.id, path="pdbs", split=TRUE, gzip=TRUE)
```

Multiple Structure Alignment
```{r}
# Align releated PDBs
pdbs <- pdbaln(files, fit = TRUE, exefile="./muscle")
```
```{r}
# Vector containing PDB codes for figure axis
ids <- basename.pdb(pdbs$id)
# Draw schematic alignment
plot(pdbs, labels=ids)
```

Use bio3d pca() function 
```{r}
# Perform PCA
pc.xray <- pca(pdbs)
plot(pc.xray)
```
```{r}
# Calculate RMSD
rd <- rmsd(pdbs)
# Structure-based clustering
hc.rd <- hclust(dist(rd))
grps.rd <- cutree(hc.rd, k=3)

plot(pc.xray, 1:2, col="grey50", bg=grps.rd, pch=21, cex=1)
```


```{r}
# Visualize first principal component
pc1 <- mktrj(pc.xray, pc=1, file="pc_1.pdb")
```

![](pc_1_pdb.png)
# Find a gene VMD file:CATF845.fwd CATF

![](findagene.png)

